---
title: "Untitled"
author: "Doug Barrows"
date: "April 28, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
# download packages

#source("https://bioconductor.org/biocLite.R")
#biocLite("ShortRead")
#biocLite("Rsubread")
#biocLite("Rsamtools")
#biocLite("BSgenome.Hsapiens.UCSC.hg19")
#biocLite("BSgenome.Hsapiens.UCSC.hg38")
#biocLite("BSgenome.Mmusculus.UCSC.mm10")
#biocLite("rtracklayer")
#biocLite("GenomicAlignments")
#biocLite("ChIPQC")
#biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
#biocLite("TxDb.Mmusculus.UCSC.mm10.knownGene")

#install.packages("devtools")
#install_github("ThomasCarroll/soGGi")

#install.packages("ggplot2")

################
# must have BSgenome index files for Rsubread in the same folder as this markdown
# if not, then put buildIndex = TRUE and it will get built in the pipeline
##############

#####if calling peaks!!!
# must open r studio fro terminal with command "open -a RStudio" for the system command to inherit the right path for macs2
# also make sure you are in python2 environment before you open Rstudio from terminal - "source activate python2"
# the pipeline will not call peaks if you do not put a sample sheet path in the command
######

fastq_path <- "./fastq_newcomp/"
sample_sheet_path <- "./sample_sheet_peakscalling.csv"

fastQ_to_bigwig_to_peaks(fastq_path = fastq_path, sample_sheet_path = sample_sheet_path, genome = "mm10", buildIndex = TRUE)

fastQ_to_bigwig_to_peaks <- function(fastq_path, mapq, buildIndex, sample_sheet_path,  bw, q, genome) {

library(ShortRead)
library(Rsubread)
library(Rsamtools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Hsapiens.UCSC.hg19)
library(BSgenome.Mmusculus.UCSC.mm10)
library(rtracklayer)
library(GenomicAlignments)
library(ggplot2)

  
if (missing(mapq)) {
  mapq <- 15
}
  
if (missing(buildIndex)) {
  buildIndex <- FALSE
}
 
if (missing(sample_sheet_path)) {
  print("No sample sheet included, no peaks will be called")
} 
  
fastq <- list.files(fastq_path)

ShortRead_QA_path <- paste0(fastq_path, "ShortRead_QA")
dir.create(ShortRead_QA_path)

BAM_path <- paste0(fastq_path, "BAM_files/")
dir.create(BAM_path)

ChIPQC_path <- paste0(fastq_path, "ChIP_QC")
dir.create(ChIPQC_path)

bigwig_path <- paste0(fastq_path, "bigwigs/")
dir.create(bigwig_path)
###########

# Now we loop through file to analyze each one

bam_files <- vector(mode = "character", length = length(fastq))

for (i in seq_along(fastq)) {
  
##########
# do some QC with shortread package
##########
  
print(paste0("ShortRead QA analysis for", fastq[i], "..."))
  
shortread_fastq <- readFastq(paste0(fastq_path, fastq[i]))
dupLogical <- srduplicated(shortread_fastq)
numberOfDups <- table(dupLogical)
perc_dups <- numberOfDups[2]/(numberOfDups[1] + numberOfDups[2])
names(perc_dups) <- "Percent_Duplicates"
out <- capture.output(perc_dups)
cat( out, file = paste0(ShortRead_QA_path, "/", fastq[i], "_duplication_perc.txt"), sep = "\n", append = TRUE)

myQA <- qa(paste0(fastq_path, fastq[i]))
myReport <- report(myQA, dest = paste0(ShortRead_QA_path, "/", fastq[i], "_ShortreadQA.html"))
######

# filter reads and write out a FastQ of our filtered reads
fqStreamer <- FastqStreamer(paste0(fastq_path, fastq[i]),
                            n=5000000)
filtered_fastq <- paste0( fastq_path, "filtered_",fastq[i])

TotalReads <- 0
TotalReadsFilt <- 0
while (length(fq <- yield(fqStreamer))>0) {
    TotalReads <- TotalReads+length(fq)
    filt1 <- fq[alphabetScore(fq) > 300 ]
    filt2 <- filt1[alphabetFrequency(sread(filt1))[,"N"] < 10]
    TotalReadsFilt <- TotalReadsFilt+length(filt2)
    writeFastq(filt2, filtered_fastq ,mode="a")
}
TotalReads
TotalReadsFilt


########## build the index for Rsubread alignment - only need to do this once

# Here we cycle through the major chromosomes and create a DNAStringSet object from the retrieved sequences.
if (buildIndex == TRUE) {
  if (i == 1) {
  print("Building index...")
  if (genome == "hg19") {
  mainChromosomes <- paste0("chr", c(1:19, "X", "Y", "M"))
  mainChrSeq <- lapply(mainChromosomes, function(x) BSgenome.Hsapiens.UCSC.hg19[[x]])
  names(mainChrSeq) <- mainChromosomes
  mainChrSeqSet <- DNAStringSet(mainChrSeq)
  
  # Now we have a DNAStringSet object we can use the writeXStringSet to create our FASTA file of sequences to align to.
  
  writeXStringSet(mainChrSeqSet,
  "BSgenome.Hsapiens.UCSC.hg19.fa")
  
  #The Rsubread package offers a faster aligner than the QuasR package although the Rsubread package. For alignment with the Rsubread package we must first build our genome index for Rsubread using the buildindex() function.The buildindex() function simply takes the parameters of our desired index name and the FASTA file to build index from.
  
  buildindex("BSgenome.Hsapiens.UCSC.hg19",
  "BSgenome.Hsapiens.UCSC.hg19.fa")
  }
  if (genome == "mm10") {
  mainChromosomes <- paste0("chr", c(1:19, "X", "Y", "M"))
  mainChrSeq <- lapply(mainChromosomes, function(x) BSgenome.Mmusculus.UCSC.mm10[[x]])
  names(mainChrSeq) <- mainChromosomes
  mainChrSeqSet <- DNAStringSet(mainChrSeq)
  
  # Now we have a DNAStringSet object we can use the writeXStringSet to create our FASTA file of sequences to align to.
  
  writeXStringSet(mainChrSeqSet,
  "BSgenome.Mmusculus.UCSC.mm10.fa")
  
  #The Rsubread package offers a faster aligner than the QuasR package although the Rsubread package. For alignment with the Rsubread package we must first build our genome index for Rsubread using the buildindex() function.The buildindex() function simply takes the parameters of our desired index name and the FASTA file to build index from.
  
  buildindex("BSgenome.Mmusculus.UCSC.mm10",
  "BSgenome.Mmusculus.UCSC.mm10.fa")
  }
}
}

###### now to the alignment

# extract the base part of the name of each fastq file so that this can be ussed as the base for output files generated below
if (grepl(".fastq", filtered_fastq)) {
  base <- strsplit(fastq[i], ".fastq", fixed = TRUE)[[1]][1]
} else if (grepl(".fq", filtered_fastq)){
  base <- strsplit(fastq[i], ".fq", fixed = TRUE)[[1]][1]
} 

#align with Rsubread

bam <- paste0(base, ".BAM")

# here I am using the filtered fastq, do we need to do this given that we filter out low quality alignments? I suppose we don't, but it might make the alignments faster if we are gettign rid of bad reads befroehand

print(paste0("Aligning ", fastq[i], "..."))

if (genome == "hg19") {
out <- capture.output(align(index = "BSgenome.Hsapiens.UCSC.hg19",
      readfile1 = filtered_fastq,
      output_file = paste0(BAM_path, bam),
      type="dna"))  # in Tom's class he had "phredOffset = 64" in this command too, but Rsubread kept throwing a warning saying it was wrong, so I'll just keep the default for now

cat( out, file = paste0(BAM_path, fastq[i], "_BAMlog.txt"), sep = "\n", append = TRUE)
}

if (genome == "mm10") {
out <- capture.output(align(index = "BSgenome.Mmusculus.UCSC.mm10",
      readfile1 = filtered_fastq,
      output_file = paste0(BAM_path, bam),
      type="dna"))  # in Tom's class he had "phredOffset = 64" in this command too, but Rsubread kept throwing a warning saying it was wrong, so I'll just keep the default for now

cat( out, file = paste0(BAM_path, fastq[i], "_BAMlog.txt"), sep = "\n", append = TRUE)
}

print(paste0("Sorting ", fastq[i], "..."))
sorted_bam_pre <- paste0( BAM_path, base, "_sorted")
sortBam(paste0(BAM_path, bam), sorted_bam_pre) # don't put ".BAM" at end of output file as it adds it anyway

sorted_bam_post <- paste0(BAM_path, base, "_sorted.BAM")
indexBam(sorted_bam_post)

file.remove(paste0(BAM_path, bam))

out <- capture.output(quickBamFlagSummary(sorted_bam_post))
cat("BamFlagSummary - Before Filtering", out, file = paste0(BAM_path, fastq[i], "_BAMlog.txt"), sep = "\n", append = TRUE)

# filter out low quality reads

print(paste0("Filtering out low quality reads and duplicates for ", fastq[i], "..."))
param <- ScanBamParam(mapqFilter = mapq, 
                       flag = scanBamFlag(isDuplicate = FALSE))
sorted_bam_post_mapq <- paste0(sorted_bam_pre, "nodup_q", as.character(mapq), ".BAM")
filterBam(file = sorted_bam_post, destination = sorted_bam_post_mapq, param = param)

bam_files[i] <- sorted_bam_post_mapq # this will eventualy be used to get the exact names of the BAM files generated to call peaks

out <- capture.output(quickBamFlagSummary(sorted_bam_post_mapq))
cat("BamFlagSummary - After Filtering", out, file = paste0(BAM_path, fastq[i], "_BAMlog.txt"), sep = "\n", append = TRUE)

# make figure showing all mapped reads to each chromosome
mappedReads <- idxstatsBam(sorted_bam_post_mapq)
TotalMapped <- sum(mappedReads[,"mapped"])
ggplot(mappedReads,aes(x=seqnames,y=mapped))+
  geom_bar(stat="identity") + ggtitle(fastq[i]) + coord_flip()
ggsave(paste0(BAM_path, fastq[i], "_chromDistr.pdf"))
  
#######
# ChIP QC
library(ChIPQC)
# using the pre-filtered BAM
print("ChIP QC for unfiltered fastq[i]...")

if (genome == "hg19") {
QCresult <- ChIPQCsample(reads = sorted_bam_post,
                         annotation="hg19",
                         blacklist = "~/Desktop/R_functions/ChIP-seq_pipeline/ChIPseq_pipeline//wgEncodeHg19ConsensusSignalArtifactRegions.bed")
out <- capture.output(QCmetrics(QCresult))
cat("ChIP QC Metrics for unfiltered", out, file = paste0(ChIPQC_path, "/", fastq[i], "_ChIPQC_output.txt"), sep = "\n", append = TRUE)
}

if (genome == "mm10") {
QCresult <- ChIPQCsample(reads = sorted_bam_post,
                         annotation="mm10",
                         blacklist = "~/Desktop/R_functions/ChIP-seq_pipeline/ChIPseq_pipeline/mm10.blacklist.bed")
out <- capture.output(QCmetrics(QCresult))
cat("ChIP QC Metrics for unfiltered", out, file = paste0(ChIPQC_path, "/", fastq[i], "_ChIPQC_output.txt"), sep = "\n", append = TRUE)
}

myFlags <- flagtagcounts(QCresult)
out <- capture.output(myFlags["DuplicateByChIPQC"]/myFlags["Mapped"])
cat( out, file = paste0(ChIPQC_path, "/", fastq[i], "_ChIPQC_output.txt"), sep = "\n", append = TRUE)

plotCC(QCresult) + ggtitle(fastq[i])
ggsave(paste0(ChIPQC_path, "/", fastq[i], "_ChIPQC_CCplot.pdf")) 

plotSSD(QCresult)+xlim(0,5) + ggtitle(fastq[i])
ggsave(paste0(ChIPQC_path, "/", fastq[i], "_ChIPQC_SSDplot.pdf"))

######

# make big wig

print(paste0("Making BigWig for ", fastq[i], "..."))
alignment <- readGAlignments(sorted_bam_post_mapq)
reads_coverage <- coverage(alignment)

export.bw(reads_coverage, con = paste0(bigwig_path, base, "_bigWig.bw"))
}

##########
# get plots over genes using soGGi
##########

library(devtools)
library(soGGi)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)

# get GRanges object of all genes

if (genome == "hg19") {
whole_gene <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
}

if (genome == "mm10") {
whole_gene <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
}

bigWigs <- list.files(paste0(fastq_path, "bigwigs"), full.names = TRUE)
whole_gene_plots <- vector(mode = "list", length = length(bigWigs))

names(whole_gene) <- NULL # needed to do this or the "percentOfRegion" setting below threw an error

for (i in seq_along(bigWigs)) {
  
  whole_gene_plots[[i]] <- regionPlot(bigWigs[i],
                        testRanges = whole_gene,
                        style = "percentOfRegion",
                        format = "bigwig")
  
}

# this will make a figure for each chip separately, need to work on code below in the chunk devoted to soGGi with either rbind, or the cconcatenation to get all in same plot. Basically it wouldn't properly concatenate in the for loop as I had it set up below. 
for(i in seq_along(whole_gene_plots)) {
  plotRegion(whole_gene_plots[[i]])
  ggsave(paste0(strsplit(metadata(whole_gene_plots[[i]])$names, ".bw", fixed = TRUE)[[1]], ".pdf"))
}


#######

# this will only run peaks if a sample sheet path is specified in the original call to the function
if (!missing(sample_sheet_path)) {
# call peaks with macs2
print("Calling Peaks...")

data <- read.delim(sample_sheet_path, header = TRUE, sep = ",")

  if (missing(bw)) {
    bw <- 300
  }
  if (missing(q)) {
    q <- 0.05
  }
  if (missing(genome)) {
    macs2_genome <- "hs"
   }
  if (genome == "hg19") {
    macs2_genome <- "hs"
   }
  if (genome == "mm10") {
    macs2_genome <- "mm"
   }

  for (i in 1:nrow(data)) {
     if (data[i,4] == 0) {
      dir.create(paste0(fastq_path, data[i,1], "_peaks/"))
      path_output <- paste0(fastq_path, data[i,1], "_peaks/")
      input <- bam_files[grep(data[i,3], bam_files)]
      peaks <- bam_files[grep(data[i,2], bam_files)]
      
      macsCommand <- paste0("macs2 callpeak", 
                   " -t ", peaks, 
                   " -c ", input, 
                   " -g ", macs2_genome,
                   " --outdir ", path_output, 
                   " -n ",  data[i,1], 
                   " --bw ", bw, " -q ", q)  
     }
     else {
       dir.create(paste0(fastq_path, data[i,1], "_broad_peaks/"))
      path_output <- paste0(fastq_path, data[i,1], "_broad_peaks/")
      input <- bam_files[grep(data[i,3], bam_files)]
      peaks <- bam_files[grep(data[i,2], bam_files)]
      
      macsCommand <- paste0("macs2 callpeak", 
                   " -t ", peaks, 
                   " -c ", input, 
                   " -g ", macs2_genome, 
                   " --outdir ", path_output, 
                   " -n ",  data[i,1], 
                   " --broad", 
                   " --bw ", bw, " -q ", q)  
     }
    system(macsCommand)
    
  #genomic annotation
    if (genome == "hg19") {
    txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
    } else if (genome == "hg38") {
    txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
    } else if (genome == "mm9") {
    txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene
    } else if (genome == "mm10") {
    txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
    }
    if (data[i,4] == 0) {
    peakAnnoList_marklist <- annotatePeak(peak = paste0(path_output, data[i,1], "_peaks.narrowPeak" ),
                                          TxDb = txdb,
                                          tssRegion = c(-5000, 3000),
                                          verbose = FALSE)
    plotAnnoBar(peakAnnoList_marklist)
    ggsave(paste0(path_output, data[i,1], "_annotation.pdf"), height = 4, width =  6)
   } else {
    peakAnnoList_marklist <- annotatePeak(peak = paste0(path_output, data[i,1], "_peaks.broadPeak" ),
                                          TxDb = txdb,
                                          tssRegion = c(-5000, 3000),
                                          verbose = FALSE)
    plotAnnoBar(peakAnnoList_marklist)
    ggsave(paste0(path_output, data[i,1], "_annotation.pdf"), height = 4, width =  6)
    
  }
}


}
```
# QA
```{r}
library(ShortRead)
library(htmlwidgets)

fastQ <- readFastq("./bladder_control_ENCFF188POD.fastq")
dupLogical <- srduplicated(fastQ)
numberOfDups <- table(dupLogical)
nonDups <- fastQ[dupLogical]

myQA <- qa("./small_fastq/sampled_ENCFF000CXH.fastq")
myReport <- report(myQA, dest = paste0(fastq_path, fastq[i], "_ShortreadQA.html"))
browseURL(myReport)
report(myReport, dest = paste0(fastq_path, fastq[i], "_ShortreadQA.html"))
saveWidget(myReport, file = paste0(fastq_path, fastq[i], "_ShortreadQA.html"))

# cant get this to work - keeps on saying permission is denied but it should be able to read and write the file. If I change permissions to let it execute it says "cannot execute binary file". 
library(fastqcr)
fastqc_path <- list.files("./fastq/", full.names = TRUE)
fastqc(fastqc.path = fastqc_path)

```
### ChIP QC

```{r}
library(ChIPQC)

# using the pre-filtered BAM
QCresult <- ChIPQCsample(reads="./filtered_bladder_control_ENCFF188POD_sorted_q15_nodup15.BAM",
                         annotation="hg19",
                         blacklist = "./wgEncodeHg19ConsensusSignalArtifactRegions.bed")
QCmetrics(QCresult)

myFlags <- flagtagcounts(QCresult)
myFlags["DuplicateByChIPQC"]/myFlags["Mapped"]

plotCC(QCresult)
plotSSD(QCresult)+xlim(0,5)

# use lapply to run on multiple BAMS
#bamsToQC <- c("Sorted_Myc_Ch12_1.bam","Sorted_Myc_Ch12_2.bam",
#             "Sorted_Myc_MEL_1.bam","Sorted_Myc_MEL_2.bam",
#             "Sorted_Input_MEL.bam","Sorted_Input_Ch12.bam")
#myQC <- bplapply(bamsToQC,ChIPQCsample,
#        annotation = "mm10",
#        blacklist = toBlkList,
#        chromosomes = paste0("chr",1:10))
#names(myQC) <- bamsToQC
#QCmetrics(myQC)
#plotCC(myQC,facetBy = "Sample")
#plotSSD(myQC)+xlim(0,5)

# We can include metadata as a data.frame where the first column is our sample names.
#myMeta <- data.frame(Sample= names(myQC),
#                     Tissue=c("Ch12","Ch12","MEL","MEL","MEL","Ch12"),
#                     Antibody=c(rep("Myc",4),rep("Input",2)))
#myMeta


```



###peak calling

```{r}
input <- c("input1","input2", "input3")
bam <- c("bam1","bam2", "bam3")
broad <- c(0,1,0)
base <- c("base1", "base2", "base3")

data <- data.frame(Base = base, BAM = bam, Input = input, Broad = broad)

#############

data <- read.delim("./sample_sheet_peakscalling.csv", header = TRUE, sep = ",")

macs2_loop <- function (data,  bw, q, genome){
  
  if (missing(bw)) {
    bw <- 150
  }
  if (missing(q)) {
    q <- 0.05
  }
    if (missing(genome)) {
    genome <- "hs"
  }

  for (i in 1:nrow(data)) {
     if (data[i,4] == 0) {
      dir.create(paste0("./", data[i,1], "_peaks/"))
      path_output <- paste0("./", data[i,1], "_peaks/")
      input <- bam_files[grep(data[i,3], bam_files)]
      peaks <- bam_files[grep(data[i,2], bam_files)]
      
      macsCommand <- paste0("macs2 callpeak", 
                   "-t ", peaks, 
                   " -c ", input, 
                   " -g ", genome,
                   " --outdir ", path_output, 
                   " -n ",  data[i,1], 
                   " --bw ", bw, " -q ", q)  
     }
     else {
       dir.create(paste0("./", data[i,1], "_broad_peaks/"))
      path_output <- paste0("./", data[i,1], "_broad_peaks/")
      input <- bam_files[grep(data[i,3], bam_files)]
      peaks <- bam_files[grep(data[i,2], bam_files)]
      
      macsCommand <- paste0("macs2 callpeak", 
                   " -t ", peaks, 
                   " -c ", input, 
                   " -g ", genome, 
                   " --outdir ", path_output, 
                   " -n ",  data[i,1], 
                   " --broad", 
                   " --bw ", bw, " -q ", q)  
     }
  }
  
  system(macsCommand)

}
  


macs2_loop(data = data)

#### must open r studio fro terminal with command "open -a RStudio" for the system command to inherit the right path
dir.create("./PeakDirectory2")
ChIP <- "./fastq_newcomp/BAM_files/filtered_K27Ac_bladder_ENCFF413QKU_sortednodup_q15.BAM"
control <- "./fastq_newcomp/BAM_files/filtered_bladder_control_ENCFF188POD_sortednodup_q15.BAM"

macsCommand <- paste0("macs2 callpeak -t ", ChIP,
                      " -c ", control,
                      " -n ", "K27ac_encodebladder",
                      " --outdir PeakDirectory2")
system(macsCommand)
system("echo $PATH")
```

### ngs plots using soggi

```{r}
library(devtools)
library(soGGi)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

# get regions of 500 bp on either side of TSS
whole_gene <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
TSS <- resize(whole_gene,500,fix = "start")
#TTS <- resize(dsds,500,fix = "end")


#### make loop to make ChIP profile object for all in big wig file and make plot for all! - TSS!

bigWigs <- list.files(paste0(fastq_path, "bigwigs"), full.names = TRUE)
TSS_plots <- vector(mode = "list", length = length(bigWigs))

for (i in seq_along(bigWigs)) {
  
  TSS_plots[i] <- regionPlot(bigWigs[i],
                        testRanges = TSS,
                        style = "point",
                        format = "bigwig")
}


for( i in seq_along(TSS_plots)) {
  if (i == 1){
    TSS_plots_cat <- TSS_plots[[i]]
  }
  else {
    TSS_plots_cat <- c(TSS_plots_cat, TSS_plots[[i]])
  }
  
}

plotRegion(TSS_plots_cat)

for( i in seq_along(TSS_plots)) {
  if (i == 1){
    TSS_plots_rbind <- TSS_plots[[i]]
  }
  else {
    TSS_plots_rbind <- rbind(TSS_plots_rbind, TSS_plots[[i]])
  }
  
}

plotRegion()

#### make loop to make ChIP profile object for all in big wig file and make plot for all! - whole gene!

bigWigs <- list.files(paste0(fastq_path, "bigwigs"), full.names = TRUE)
whole_gene_plots <- vector(mode = "list", length = length(bigWigs))

names(whole_gene) <- NULL # needed to do this or the "percentOfRegion" setting below threw an error

for (i in seq_along(bigWigs)) {
  
  whole_gene_plots[[i]] <- regionPlot(bigWigs[i],
                        testRanges = whole_gene,
                        style = "percentOfRegion",
                        format = "bigwig")
  
}

# this will make a figure for each chip separately, need to work on code below with either rbind, or the cconcatenation to get all in same plot
for(i in seq_along(whole_gene_plots)) {
  plotRegion(whole_gene_plots[[i]])
  ggsave(paste0(strsplit(metadata(whole_gene_plots[[i]])$names, ".bw", fixed = TRUE)[[1]], ".pdf"))
}

# this only works for 1 or two chips, after the second it has trouble concatenating them properly. If you manually concatenate them, it works
for( i in seq_along(whole_gene_plots)) {
  if (i == 1){
    whole_gene_plots_cat <- whole_gene_plots[[i]]
  }
  else {
    whole_gene_plots_cat <- c(whole_gene_plots_cat, whole_gene_plots[[i]])
  }
  
}

plotRegion(whole_gene_plots_cat)

for( i in seq_along(whole_gene_plots)) {
  if (i == 1){
    whole_gene_plots_rbind <- whole_gene_plots[[i]]
  }
  else {
    whole_gene_plots_rbind <- rbind(whole_gene_plots_rbind, whole_gene_plots[[i]])
  }
  
}

plotRegion()
```


